modular/executor/execute_task.go

```diff
@@ -162,108 +163,118 @@ func (e *ExecuteModular) HandleReceivePieceTask(
 	ctx context.Context,
 	task coretask.GCObjectTask) {
-	var err error
+	var (
+		err                error
+		toGCObjects        []*types.Object
+		currentGCBlockID   uint64
+		currentGCObjectID  uint64
+		responseEndBlockID uint64
+		storageParams      *storagetypes.Params
+		gcObjectNumber     int
+		tryAgainLater      bool
+		taskIsCanceled     bool
+		hasNoObject        bool
+		isSucceed          bool
+	)
+
+	reportProgress := func() bool {
+		reportErr := e.ReportTask(ctx, task)
+		log.CtxDebugw(ctx, "gc object task report progress", "task_info", task.Info(), "error", reportErr)
+		return errors.Is(reportErr, manager.ErrCanceledTask)
+	}
+
 	defer func() {
-		task.SetError(err)
+		if err == nil && (isSucceed || hasNoObject) { // succeed
+			task.SetCurrentBlockNumber(task.GetEndBlockNumber() + 1)
+			reportProgress()
+		} else { // failed
+			task.SetError(err)
+			reportProgress()
+		}
+		log.CtxDebugw(ctx, "gc object task",
+			"task_info", task.Info(), "is_succeed", isSucceed,
+			"response_end_block_id", responseEndBlockID, "to_gc_object_number", len(toGCObjects),
+			"has_gc_object_number", gcObjectNumber, "try_again_later", tryAgainLater,
+			"task_is_canceled", taskIsCanceled, "has_no_object", hasNoObject, "error", err)
 	}()
-
-	objects, endBlockNumber, err := e.baseApp.GfSpClient().ListDeletedObjectsByBlockNumberRange(
-		ctx, e.baseApp.OperateAddress(), task.GetStartBlockNumber(),
-		task.GetEndBlockNumber(), true)
-	if err != nil {
+	// TODO: refine it to support multi version
+	if storageParams, err = e.baseApp.Consensus().QueryStorageParams(context.Background()); err != nil {
+		log.Errorw("failed to query storage params", "task_info", task.Info(), "error", err)
 		return
 	}
-	if len(objects) == 0 {
-		task.SetCurrentBlockNumber(task.GetEndBlockNumber() + 1)
+	if toGCObjects, responseEndBlockID, err = e.baseApp.GfSpClient().ListDeletedObjectsByBlockNumberRange(
+		ctx, e.baseApp.OperateAddress(), task.GetStartBlockNumber(),
+		task.GetEndBlockNumber(), true); err != nil {
+		log.CtxErrorw(ctx, "failed to query deleted object list", "task_info", task.Info(), "error", err)
 		return
 	}
-
-	cancel := func() bool {
-		return errors.Is(e.ReportTask(ctx, task), manager.ErrCanceledTask)
+	if responseEndBlockID < task.GetStartBlockNumber() || responseEndBlockID < task.GetEndBlockNumber() {
+		// metadata is not latest, try again later.
+		tryAgainLater = true
+		log.CtxInfow(ctx, "metadata is not latest, try again later",
+			"response_end_block_id", responseEndBlockID, "task_info", task.Info())
+		time.Sleep(10 * time.Second)
+		return
 	}
-
-	var (
-		objectInfo       *storagetypes.ObjectInfo
-		deletingBlock    uint64
-		deletingObjectID uint64
-		deletingIdx      int
-		pieceKey         string
-		segmentCount     uint32
-		params           *storagetypes.Params
-	)
-	params, err = e.baseApp.GfSpDB().GetStorageParams()
-	if err != nil {
-		log.CtxErrorw(ctx, "failed to get params", "error", err)
+	if len(toGCObjects) == 0 {
```

Comment:
what is the meaning of `toGCObjects`? The objects that waiting to GC?

Reply:
Has Updated to `waitingGCObjects`.